unit_tests/log_transfer/test_main_async_log_page_fetcher.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#define CATCH_CONFIG_MAIN
+
+#include "catch2/catch.hpp"
+#include "log_page_async_fetcher.hpp"
+#include "log_reader.hpp"
+
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+
+struct page_id_requirement
+{
+  bool require_log_page_valid;
+  bool is_page_received;
+};
+struct log_page_fetcher_test_data
+{
+  std::unordered_map<LOG_PAGEID, page_id_requirement> page_ids_requested;
+  std::mutex map_mutex;
+} g_log_page_fetcher_test_data;
+
+class test_env
+{
+  public:
+    test_env (bool require_log_page_valid, std::vector<LOG_PAGEID> log_pageids);
+    ~test_env ();
+
+    void run_test ();
+
+  private:
+    void on_receive_log_page (LOG_PAGEID page_id, const LOG_PAGE *, int error_code);
+
+  private:
+    std::vector<LOG_PAGEID> m_log_pageids;
+    cublog::async_page_fetcher m_async_page_fetcher;
+};
+
+void do_test (test_env &env)
+{
+  env.run_test ();
+}
+
+TEST_CASE ("Test with a valid log page returned", "")
+{
+  THREAD_ENTRY *thread_p = NULL;
+  cubthread::initialize (thread_p);
+  cubthread::initialize_thread_entries ();
+
+  std::vector<LOG_PAGEID> page_ids { 1, 2, 3 };
+  test_env env (true, page_ids);
+  do_test (env);
+}
+
+TEST_CASE ("Test with an invalid log page returned", "")
+{
+  std::vector<LOG_PAGEID> page_ids { 4, 5, 6 };
+  test_env env (false, page_ids);
+  do_test (env);
+}
+
+TEST_CASE ("Test with a very big number of pages", "")
+{
+  std::vector<LOG_PAGEID> page_ids;
+  for (auto i = 1; i < 10001; ++i)
```

Comment:
That's blasphemy :))
```suggestion
  for (auto i = 0; i < 10000; ++i)
```